### ISSUE TITLE
Push replication failing with large data sets

### ISSUE THREAD

**Issue by swilks** [couchbase/couchbase-lite-android#898]
Hey,

I'm developing an Android app in Kenya with a user base is that is often rural and doesn't have consistent access to the internet. We need to back up our users' data, and have just started to use couchbase lite for Android. Because of their poor connectivity they often go long periods of time without uploading any data to the server so push replications are frequently pretty huge. We need push to work with any reasonably large set of documents that take up a reasonably large amount of space. Luckily we don't need to sync any images are large files, just documents that are generally pretty small. 

I haven't plugged in a real dataset to test yet, and am just testing with one type of document with randomly generated data to make sure I can roll out a scalable solution before we drop that app into real-world conditions. I'm currently testing on an Alcatel One Touch (which is a pretty sub-standard tablet, but it's the kind of thing are users are likely to use, so it's a good test device) with ~65MB worth of documents (25,000 docs at ~2.5kb each). I flush the server before running the tests, so it's purely a push replication.

By the time the ChangeEvent has registered about 24,000 changes (it hasn't started uploading at this point) it starts generating Out of Memory errors.

First of all, I'm not sure why the upload isn't incremental in this case? When I test with smaller data sets the upload starts before the CBLite has registered all of changed items. For example, when I run the push replication with 750 items, the upload starts after getChangeCount() returns 500 items, and that number is smaller (proportionally) with smaller data sets. 

Second, is there anything I can do to change this behaviour? I have some hacky solutions in mind to break up the upload, but I don't want to implement them if I don't have to.

Here are the OOM errors I'm getting: 
```
06-09 19:31:20.321 6978-7368/org.mc.mc E/Sync: RemoteRequest.run() exception: %s
java.lang.OutOfMemoryError
      at com.fasterxml.jackson.core.util.BufferRecycler.calloc(BufferRecycler.java:156)
      at com.fasterxml.jackson.core.util.BufferRecycler.allocCharBuffer(BufferRecycler.java:124)
      at com.fasterxml.jackson.core.util.BufferRecycler.allocCharBuffer(BufferRecycler.java:114)
      at com.fasterxml.jackson.core.io.IOContext.allocConcatBuffer(IOContext.java:186)
      at com.fasterxml.jackson.core.json.UTF8JsonGenerator.<init>(UTF8JsonGenerator.java:126)
      at com.fasterxml.jackson.core.JsonFactory._createUTF8Generator(JsonFactory.java:1284)
      at com.fasterxml.jackson.core.JsonFactory.createGenerator(JsonFactory.java:1016)
      at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:2801)
      at com.couchbase.lite.support.RemoteRequest.setBody(RemoteRequest.java:326)
      at com.couchbase.lite.support.RemoteRequest.run(RemoteRequest.java:96)
      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
      at java.util.concurrent.FutureTask.run(FutureTask.java:237)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
      at java.lang.Thread.run(Thread.java:841)
06-09 19:31:23.645 6978-7371/org.mc.mc E/dalvikvm-heap: Out of memory on a 7817-byte allocation.
06-09 19:31:23.652 6978-7371/org.mc.mc E/Sync: RemoteRequest.run() exception: %s
java.lang.OutOfMemoryError
      at com.fasterxml.jackson.core.util.ByteArrayBuilder.toByteArray(ByteArrayBuilder.java:118)
      at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:2807)
      at com.couchbase.lite.support.RemoteRequest.setBody(RemoteRequest.java:326)
      at com.couchbase.lite.support.RemoteRequest.run(RemoteRequest.java:96)
      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
      at java.util.concurrent.FutureTask.run(FutureTask.java:237)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
      at java.lang.Thread.run(Thread.java:841)
```

and here's my sync function:
```java
final Replication pull = database.createPullReplication(createSyncURL(false));
        final Replication push = database.createPushReplication(createSyncURL(false));

        pull.setContinuous(false);
        pull.addChangeListener((new Replication.ChangeListener() {
            @Override
            public void changed(Replication.ChangeEvent event) {
                processSyncChangeEvent(event, pull, push, notifyManager, builder, appContext
                        .getString(R.string.downloaded_items), appContext.getString(R.string
                        .prep_downloaded_items), appContext);
            }
        }));
        if (pullListener != null) {
            pull.addChangeListener(pullListener);
        }

        push.setContinuous(false);

        push.addChangeListener(new Replication.ChangeListener() {
            @Override
            public void changed(Replication.ChangeEvent event) {
                processSyncChangeEvent(event, push, pull, notifyManager, builder, appContext
                        .getString(R.string.uploaded_items), appContext.getString(R.string
                        .prep_uploaded_items), appContext);

            }
        });
        if (pushListener != null) {
            push.addChangeListener(pushListener);
        }
        pull.start();
        push.start();
```

**Comment by hideki** [comment]
Hi @swilks,

Thank you very much for your report.

- What version of Couchbase Lite Android are you using? v1.2.1? v1.2.0?...
- What is server? Sync Gateway or CouchDB? and what is its version?
- Can you provide us sample data?

Thanks!

**Comment by hideki** [comment]
Hi @sethrosetter,

> I'm currently testing on an Alcatel One Touch (which is a pretty sub-standard tablet, but it's the kind of thing are users are likely to use, so it's a good test device) with ~65MB worth of documents (25,000 docs at ~2.5kb each)

Can you implement functional test for single push replication with 25,000 docs which average doc size is 2.5KB.

**Comment by swilks** [comment]
Hi @hideki,

- using v1.2.1
- sync gateway v. 1.2.1 on AWS ec2 t2 micro (which I know is too small for Couchbase server, but should work for testing right? It hasn't been overloaded yet). Couchbase server community edition 4.0.0
-I'm on my phone right now and don't have access to my computer, but the docs are 4

**Comment by swilks** [comment]
Sorry! Finger slipped.

- docs are 2 string fields with length 10-20 chars, 2 with length 2-5chars, and one with length ~1500 chars (to inflate the size of each doc for testing). 2 ints, two doubles

**Comment by hideki** [comment]
@swilks Thank you for additional information. We will try to reproduce the issue.

**Comment by swilks** [comment]
Sorry for the slow reply (time difference issues).

The model number is 9005x
8gb storage, 1gb ram
ARMv7 Processor rev3 (v7l) - 1.3ghz, pretty sure it's dual core

**Comment by swilks** [comment]
It's the Alcatel Pixi 8, 8gb version

**Comment by swilks** [comment]
Hey, I'm wondering what the status of this is? On reviewable it looks like it was marked as fixed? Is there any way I can incorporate these changes, or do I have to wait for the next release?

**Comment by hideki** [comment]
This ticket is still in Incoming queue. It means this is not fixed yet.

**Comment by hideki** [comment]
We removed pause logic, which prevent to process many requests in case of many documents are created or updated, because the implementation was not thread-safe, and it caused dead-lock.
https://github.com/couchbase/couchbase-lite-java-core/commit/d990d7d4a4a769352d35117c15c8d8f6e6f148a4#diff-f7bd4a59e65ba5e7a9ab2b15276b81f5

We need to implement pause logic for push replicator with thread-safty (no dead-lock.)

**Comment by swilks** [comment]
Great, thanks for the update. Any idea what the timeline for the fix will be? I can attempt a hackey solution in the meantime, but if this is a high-priority bug is that likely going to be fixed soon I won't bother

**Comment by hideki** [comment]
I could reproduce this OOM issue.
https://gist.github.com/hideki/5fce3fcb0a1752494a267842efde592d

**Comment by hideki** [comment]
Hi @swilks,

We fixed this issue. If you have time, could you please verify the fix with using dev build for 1.3.0?
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.0-4/

Instruction: how to setup dependency to Couchbase custom maven.
https://github.com/couchbase/couchbase-lite-android#using-latest-official-release

Thanks!

**Comment by swilks** [comment]
Hey @hideki, sure! I'm not sure when I'll get time as we're running up against some deadlines, but I can let you know by early next week at the latest

**Comment by swilks** [comment]
Hey @hideki sorry for the slow response. I've tested the change, but ran into server-side problems that are preventing me from verifying it 100%.

I loose connection to the server after about 15000 documents, but push behaviour seems correct. I was hoping I could ask your opinion on what the likely problem is. I'm using a t2.micro instance on AWS which is pretty small (~1gb ram)  and I'm guessing that my instance is just too small to support anything remotely intensive. Does that sound correct? (I posted the full issue here: https://forums.couchbase.com/t/couchbase-server-down-after-large-push-replication-from-cblite-android-on-ec2-t2-micro/9083)

Thanks!

**Comment by hideki** [comment]
server side minimum requirements included here: http://developer.couchbase.com/documentation/mobile/current/develop/guides/sync-gateway/deployment/in-production/index.html

- Couchbase Server -Quad-core/16GB RAM	
- Sync Gateway - Quad-core/4GB RAM

To isolate if the issue you are encountering is hosting server issue or bug in sync gateway or couchbase server, I recommend to test your app with Couchbase recommendation environment at least once.

**Comment by swilks** [comment]
@hideki thanks! I upgraded my instance push replications are working fine!

Quick follow-up. After I push I'm resetting my local database to attempt to pull the same items back from the server and the pull replication is is going through the first 50 items and then "successfully" completing. Any ideas what could be causing this? My pull replication code is essentially the same, and is stopping with the status STOPPED after 50 items are pulled. Push replications are working as expected, and the only change I've made is to change my sync class from static to a singleton. 

Could it be problem with sync gateway?
